packageurl_test.go

```diff
@@ -119,150 +116,198 @@ func (t TestFixture) Qualifiers() packageurl.Qualifiers {
+			t.Logf("%s: incorrect namespace: wanted: '%s', got '%s'", tc.Description, expected.Namespace, p.Namespace)
+			t.Fail()
+		}
+		if p.Name != expected.Name {
+			t.Logf("%s: incorrect name: wanted: '%s', got '%s'", tc.Description, expected.Name, p.Name)
+			t.Fail()
+		}
+		if p.Version != expected.Version {
+			t.Logf("%s: incorrect version: wanted: '%s', got '%s'", tc.Description, expected.Version, p.Version)
+			t.Fail()
+		}
+		want := expected.Qualifiers()
+		sort.Slice(want, func(i, j int) bool {
+			return want[i].Key < want[j].Key
+		})
+		got := p.Qualifiers
+		sort.Slice(got, func(i, j int) bool {
+			return got[i].Key < got[j].Key
+		})
+		if !reflect.DeepEqual(want, got) {
+			t.Logf("%s: incorrect qualifiers: wanted: '%#v', got '%#v'", tc.Description, want, p.Qualifiers)
+			t.Fail()
+		}
+
+		if p.Subpath != expected.Subpath {
+			t.Logf("%s: incorrect subpath: wanted: '%s', got '%s'", tc.Description, expected.Subpath, p.Subpath)
+			t.Fail()
 		}
-		instance := packageurl.NewPackageURL(
-			tc.PackageType, tc.Namespace, tc.Name, tc.Version,
-			// Use QualifiersFromMap so that the qualifiers have a defined order, which is needed for string comparisons
-			packageurl.QualifiersFromMap(tc.Qualifiers().Map()), tc.Subpath)
-		result := instance.ToString()
-
-		// NOTE: We create a purl with ToString and then load into a PackageURL
-		//       because qualifiers may not be in any order. By reparsing back
-		//       we can ensure the data transfers between string and instance form.
-		canonical, _ := packageurl.FromString(tc.CanonicalPurl)
-		toTest, _ := packageurl.FromString(result)
-		// If the two results don't equal then the ToString failed
-		if !reflect.DeepEqual(toTest, canonical) {
-			t.Logf("%s failed: %s != %s", tc.Description, result, tc.CanonicalPurl)
+	} else {
+		// Invalid cases
+		if err == nil {
+			t.Logf("%s did not fail and returned %#v", tc.Description, p)
 			t.Fail()
 		}
 	}
+
 }
 
-// TestStringer verifies that the Stringer implementation produces results
-// equivalent with the ToString method.
-func TestStringer(t *testing.T) {
-	// Read the json file
-	data, err := os.ReadFile("testdata/test-suite-data.json")
-	if err != nil {
-		t.Fatal(err)
+func buildTest(tc TestFixture, t *testing.T) {
+	input := tc.Input.PurlComponent
+	instance := packageurl.NewPackageURL(
+		input.PackageType, input.Namespace, input.Name, input.Version,
+		// Use QualifiersFromMap so that the qualifiers have a defined order, which is needed for string comparisons
+		packageurl.QualifiersFromMap(input.Qualifiers().Map()), input.Subpath)
+	result := instance.ToString()
+	canonicalExpectedPurl := tc.ExpectedOutput.Purl
+
+	if tc.ExpectedFailure == false {
+		if result != *canonicalExpectedPurl {
+			t.Logf("%s: '%s' test failed: wanted: '%s', got '%s'", tc.Description, tc.TestType, *canonicalExpectedPurl, result)
+			t.Fail()
+		}
+	} else {
+		t.Logf("%s did not fail and returned %#v", tc.Description, instance)
+		t.Fail()
```

Comment:
I found the error scenarios for the `build` test cases a bit tricky. The `ToString` and the more Go-idiomatic `String` methods don't have an error in their method signatures so calling that alone will never produce an error. It seems like these tests suggest that the library is expected to fail when generating a string representation from a struct that contains an invalid purl. That, to me, sounds like a `PackageURL.Validate` call or such. But I think the closest is `PackageURL.Normalize` which I think is what does validation at the moment.

```suggestion
	if tc.ExpectedFailure {
		// String()/ToString() signature won't error so the only reasonable thing is to check this here.
		err := instance.Normalize()
		if err == nil {
			t.Logf("'%s' did not fail for %#v", tc.Description, instance)
			t.Fail()
		}
	} else {
		if result != *canonicalExpectedPurl {
			t.Logf("%s: '%s' test failed: wanted: '%s', got '%s'", tc.Description, tc.TestType, *canonicalExpectedPurl, result)
			t.Fail()
		}
	}
```